### ISSUE TITLE
Address field variation analysis

### ISSUE THREAD

**Issue by floresj** [truetandem/e-QIP-prototype#1344]
The following is a list of fields that have address information that deviate from our typical Address component. Currently, our address component enforces that a user enter all information for either domestic, international or APO addresses. We started out with using that component in an effort to normalize address information ingestion. However, this is no longer the case and we must now account for location edge cases. 

A couple of more thoughts that are not expressed with this list
- If a question asks for a full address but doesn't specify APO, should we not give the use the option to enter APO? For instance, there are many questions that follow this pattern: `Street address and city & State and Zip Code or Country` but do not mention anything about APO/FPO addresses.

## tl;dr
- City, State, Country
- Street, City, State, Zip
- City, State
- City, Country
- City, County, State or Country
- Street, City, Country
- City
- Country

### Foreign born to us parents in a foreign country

-  **Provide the place of issuance** - City, State, Country
![selection_133](https://cloud.githubusercontent.com/assets/5938731/26634184/cf049ec6-45e3-11e7-8fda-917ac83391da.png)

### Citizenship Naturalized US Citizen

- **Provide the location of entry into the U.S.** - Street, City, State, Zip
![selection_134](https://cloud.githubusercontent.com/assets/5938731/26634255/0b079a54-45e4-11e7-8b2a-1d0cc026f3a7.png)

- **Provide the address of the court that  issued the certificate of naturalization** - Street, City, State, Zip
![selection_135](https://cloud.githubusercontent.com/assets/5938731/26634270/1678f266-45e4-11e7-8477-872516541b56.png)

### Citizenship Not a US Citizen
- **Provide your place of entry in the US** - City, State
![selection_136](https://cloud.githubusercontent.com/assets/5938731/26634335/4bd9b99a-45e4-11e7-996a-59e9c31b6a1c.png)

### Foreign Passport (or Identify Card)
- **Provide the place the passport (or identity card) was issued** - City, Country

### If in A marriage, Civil Union, or Domestic Partnership or separated
- **Provide location** - City, County, State or Country
![selection_137](https://cloud.githubusercontent.com/assets/5938731/26634781/a8b0a146-45e5-11e7-902c-fda23e4ac83a.png)

- **If legally separated, provide the location of the record** - City, State and Zipcode or Country
![selection_138](https://cloud.githubusercontent.com/assets/5938731/26634804/b8e69fac-45e5-11e7-8bdd-133ef419ca00.png)

### If Widowed, Divorced/Dissolved, or Annulled
- **Provide the location** - City, State or Country
- **Provide where the record of divorce/dissolution or annulment is located** - City, State and Zipcode or Country
![selection_139](https://cloud.githubusercontent.com/assets/5938731/26634903/0b249f94-45e6-11e7-8c0a-44698b6a1cc0.png)

## Section 18 - Relatives

- **Provide your relatives place of birth** - City, State, Country
- **Provide the address of the court that issued the Certificate of Naturalization** - Street Address, City, State and Zip Code
- **Provide the address of current employer, or provide the address of their most recent employer if not currently employed** - Street Address, City
  -  **Provide Country if outside the United States; otherwise, provide State and Zip Code** - State, Zipcode, Country
![image](https://cloud.githubusercontent.com/assets/5938731/26635203/06459e64-45e7-11e7-8798-47c7e76eee8e.png)

### If Yes to Having Foreign Real Estate
- **Provide the location/address of property** - Street, City, Country

### If Yes to Offered Job
- **Provide the location where this occurred** - City, State and Zip Code or Country

### If Yes to Attending Foreign Conferences
- **Provide the city where the event was held** - City
- **Provide the country where the event was held** - Country

### If Yes to Seeking Credit Counseling
- **Provide the location of the credit counseling organization** - City, State

cc: @bmallred @Jay-Neighbours-I72

**Comment by bmallred** [comment]
So just a few deviations :smile:

**Comment by floresj** [comment]
Just some info on my prgress on normalizing location information to handle the different variations. Currently in the [location-components](https://github.com/truetandem/e-QIP-prototype/tree/location-components) branch, I have added a [Location.jsx.](https://github.com/truetandem/e-QIP-prototype/blob/location-components/src/components/Form/Location/Location.jsx)   In order to use it, you pass in the fields that should be rendered. For instance, this would render state and city fields in that order.

```jsx
<Location name="Location"
  fields={['state', 'city']}
  className="credit-location"
  bind={true}
  statePlaceholder="Please enter state or territory within the United States"
  cityPlaceholder="Please enter the city"
  ... stuff removed for brevity
/>
```

Now, for more complicated scenarios, you can simply define a key you want to hit on and define your own case. For instance, say we want to render the birthplace component. We can match on `fields={['birthPlace']}` which can delegate to the `BirthPlace` component.

```jsx
const fields = this.props.fields.map(field => {
      switch (field) {
        case 'fullAddress':
          return (
            <Address name="fulladdress"
              {...this.props}
              onUpdate={this.updateFullAddress}
              onError={this.props.onError}
            />
          )
        case 'birthPlace':
          return (
            <BirthPlace
              {...this.props}
              onUpdate={this.updateBirthPlace}
              onError={this.props.onError}
            />
          )
        case 'street':
          return (
            <Street name="address"
              key={field}
              className="mailing"
              label={i18n.t('address.us.street.label')}
              placeholder={i18n.t('address.us.street.placeholder')}
              value={this.props.address}
              onChange={this.updateStreet}
              onError={this.props.onError}
              onFocus={this.props.onFocus}
              onBlur={this.handleBlur}
            />
          )
        case 'city':
            //... render city
        case 'county':
            //... render county
       // ... other individual fields
})
```

Now, regardless of what fields are rendered, we maintain the same structure for the data being sent up and passed in.

```javascript
{
        address: this.props.address,
        city: this.props.city,
        zipcode: this.props.zipcode,
        state: this.props.state,
        country: this.props.country,
        // For the toggle on birthplace
        domestic: this.props.domestic,
        fields: this.props.fields
}
```

Then with this, we can also just use a single [location validator](https://github.com/truetandem/e-QIP-prototype/blob/location-components/src/validators/location.js) to handle all those cases.

As far as geocoding, I was thinking of allowing a `geocode={true}` prop to be passed in that would tell the Location component to execute a geocode request when the fields are populated.

cc: @bmallred

**Comment by bmallred** [comment]
- Do we still need the `domestic` property? If `country` is used then it could simply become a function.
 - Also if we're going through this iteration (which I think is a good idea) mind changing `address` to `street` or something else to reduce stuttering?
 - Is the point of returning `fields` in the structure for validation?

**Comment by floresj** [comment]
- The `domestic` prop is there for the Yes/No toggle branch thingy. We could do away with it and use the value of `null` on country to handle the third state where nothing is selected.
![selection_215](https://user-images.githubusercontent.com/5938731/27333136-dcb925b4-5592-11e7-88db-7d75f168072b.png)
- Yessssssss I'm glad you said that.
- No, that was just for convenience so I can then just render `{ fields }`

**Comment by bmallred** [comment]
I was thinking if we could utilize country to determine button selection it may be ideal. Then we wouldn't have to store layout specific properties like `domestic` or `addressType`

**Comment by floresj** [comment]
Instead of 

```jsx
<Location name="Location"
  fields={['state', 'city']}
  className="credit-location"
  bind={true}
  statePlaceholder="Please enter state or territory within the United States"
  cityPlaceholder="Please enter the city"
  ... stuff removed for brevity
/>```

Something like this:

```jsx
<Location name="Location"
  fields="StateCity"
  className="credit-location"
  bind={true}
  statePlaceholder="Please enter state or territory within the United States"
  cityPlaceholder="Please enter the city"
  ... stuff removed for brevity
/>```

For Yes/No domestic/international type fields
```jsx
        <Location name="ssn"
             toggle={true}
             domesticFields={['state', 'city', 'county']}
             internationalFields={['city', 'country']}
             {...this.props.value}
             onUpdate={this.props.onUpdate}
             onError={this.handleError}
             />
```

use this instead
```jsx
        <Location name="ssn"
             fields="DomesticStateCityCountyInternationalCityCountry"
             domesticFields={['state', 'city', 'county']}
             internationalFields={['city', 'country']}
             {...this.props.value}
             onUpdate={this.props.onUpdate}
             onError={this.handleError}
             />
```

**Comment by floresj** [comment]
Ok, this is definitely looking better after discussing with your. Here's one example with just rendering State and City fields with custom state/city placeholders (Will be localized)

```jsx
              <Location name="Location"
                          layout={Location.STATE_CITY}
                          className="credit-location"
                          bind={true}
                          statePlaceholder="Please enter state or territory within the United States"
                          cityPlaceholder="Please enter the city"
                          />
```

Here's one for Birthplace

```jsx
              <Location name="Location"
                          layout={Location.BIRTHPLACE}
                          className="credit-location"
                          bind={true}
                          />
```
cc: @bmallred

**Comment by floresj** [comment]
Just an update on how validator test data may look like:

```js
      {
        state: {
          HasCreditCounseling: 'Yes',
          List: [
            {
              Explanation: {
                value: 'Completely forgot'
              },
              Location: {
                city: 'Arlington',
                state: 'Virginia',
                country: 'United States',
                layout: Location.STATE_CITY
              },
            }
          ],
          ListBranch: 'No'
        },
        expected: true
      }
```

Just showing use of constant when mocking info.

**Comment by floresj** [comment]
Tracking issues from this sprint that required address/location updates 
- #1420

**Comment by bmallred** [comment]
@floresj good to close?

**Comment by floresj** [comment]
@bmallred Yessir!
